collectors/ebpf.plugin/ebpf_oomkill.c

```diff
@@ -0,0 +1,198 @@
+            bpf_link__destroy(probe_links[i]);
+            i++;
+        }
+        bpf_object__close(objects);
+    }
+}
+
+static void oomkill_write_data()
+{
+    // the first `i` entries of `keys` will contain the currently active PIDs
+    // in the eBPF map.
+    uint32_t i = 0;
+    int32_t keys[NETDATA_OOMKILL_MAX_ENTRIES] = {0};
+
+    uint32_t curr_key = 0;
+    uint32_t key = 0;
+    int mapfd = oomkill_maps[OOMKILL_MAP_KILLCNT].map_fd;
+    while (bpf_map_get_next_key(mapfd, &curr_key, &key) == 0) {
+        curr_key = key;
+
+        keys[i] = key;
+        i += 1;
+
+        // delete this key now that we've recorded its existence. there's no
+        // race here, as the same PID will only get OOM killed once.
+        int test = bpf_map_delete_elem(mapfd, &key);
+        if (unlikely(test < 0)) {
+            // since there's only 1 thread doing these deletions, it should be
+            // impossible to get this condition.
+            error("key unexpectedly not available for deletion.");
+        }
+    }
+
+    // for each app, see if it was OOM killed. record as 1 if so otherwise 0.
+    struct target *w;
+    for (w = apps_groups_root_target; w != NULL; w = w->next) {
+        if (likely(w->exposed && w->processes)) {
+            bool was_oomkilled = false;
+            struct pid_on_target *pids = w->root_pid;
+            while (pids) {
+                uint32_t j;
+                for (j = 0; j < i; j++) {
+                    if (pids->pid == keys[j]) {
+                        was_oomkilled = true;
+                        goto write_dim;
+                    }
+                }
+                pids = pids->next;
+            }
+
+        write_dim:;
+            write_chart_dimension(w->name, was_oomkilled);
+        }
+    }
+}
+
+/**
+* Main loop for this collector.
+*/
+static void oomkill_collector(ebpf_module_t *em)
+{
+    UNUSED(em);
+
+    // loop and read until ebpf plugin is closed.
+    while (!close_ebpf_plugin) {
+        pthread_mutex_lock(&collect_data_mutex);
+        pthread_cond_wait(&collect_data_cond_var, &collect_data_mutex);
+        pthread_mutex_lock(&lock);
+
+        // write everything from the ebpf map.
+        write_begin_chart(NETDATA_EBPF_APPS_GROUP, "oomkills");
+        oomkill_write_data();
+        write_end_chart();
+
+        pthread_mutex_unlock(&lock);
+        pthread_mutex_unlock(&collect_data_mutex);
+    }
+}
+
+/**
+ * Create apps charts
+ *
+ * Call ebpf_create_chart to create the charts on apps submenu.
+ *
+ * @param em a pointer to the structure with the default values.
+ */
+void ebpf_oomkill_create_apps_charts(struct ebpf_module *em, void *ptr)
+{
+    UNUSED(em);
+
+    struct target *root = ptr;
+    ebpf_create_charts_on_apps("oomkills",
+                               "OOM kills",
+                               EBPF_COMMON_DIMENSION_KILLS,
+                               "mem",
+                               NETDATA_EBPF_CHART_TYPE_LINE,
```

Comment:
Usually, we use "stacked" for apps charts when the sum of dimensions makes sense.

Reply:
:+1:

Reply:
https://github.com/netdata/netdata/pull/11470/commits/40c2ddbdf57c0184eff3cc4625202c37654106e6